proto/structured.proto

```diff
@@ -104,3 +105,224 @@ message CreateTableResponse {
+  // operations. The default is CONSISTENT. This value is ignored for
+  // write operations.
+  optional ReadConsistencyType read_consistency = 10 [(gogoproto.nullable) = false];
+}
+
+// TableResponseHeader is returned with every structured API response.
+message TableResponseHeader {
+  // Error is non-nil if an error occurred.
+  optional Error error = 1;
+  // Timestamp specifies time at which read or write actually was
+  // performed. In the case of both reads and writes, if the timestamp
+  // supplied to the request was 0, the wall time of the node
+  // servicing the request will be set here. Additionally, in the case
+  // of writes, this value may be increased from the timestamp passed
+  // with the RequestHeader if the key being written was either read
+  // or written more recently.
+  //
+  // TODO(vivek): Is this only relevant in the non-transactional case?
+  optional Timestamp timestamp = 2 [(gogoproto.nullable) = false];
+  // Transaction is non-nil if the request specified a non-nil
+  // transaction. The transaction timestamp and/or priority may have
+  // been updated, depending on the outcome of the request.
+  optional Transaction txn = 3;
+}
+
+// A TableKey represents a key into a table index.
+message TableKey {
+  optional string table_name = 1;
+  // A column key component. Multiple ColumnKeys together form a key.
+  message ColumnKey {
+    optional string column_name = 1;
+    optional bytes key = 2; 
+  }
+  // The different key components from each column.
+  repeated ColumnKey columns = 2;
+}
+
+// A TableRow represents a set of column values within a table row. 
+// When multiple rows are returned in a response (Scan), only the
+// the first row Cells contain column_names.
+message TableRow {
+  message Cell {
+    optional string column_name = 1;
+    optional Value value = 2;
+  }
+  repeated Cell cells = 1;
+}
+
+// A GetTableRowRequest is used to read a table row.
+message GetTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableKey key = 2 [(gogoproto.nullable) = false];
+  // get specified columns from the row.
+  repeated string column_names = 3;
+}
+
+// A GetTableRowResponse returns a row.
+// If the key doesn't exist, it returns an empty row.
+message GetTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableRow row = 2 [(gogoproto.nullable) = false];
+}
+
+// A PutTableRowRequest inserts/updates a row in the table. To write
+// an empty value in a cell specify an empty Bytes in the Value. To Delete
+// a cell, specify a cell with a column_name with no value.
+message PutTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableKey key = 2 [(gogoproto.nullable) = false];
+  optional TableRow row = 3 [(gogoproto.nullable) = false];
```

Comment:
I wonder if `key` and `row` should be merged. Separating them requires the client to know which columns are part of the primary key. Why not just stick all of the columns into the `row`? This comment applies elsewhere.